File: frontend/src/schemas/counselling.ts
/**
 * Zod schemas — Counselling module.
 */
import { z } from 'zod';
import { referralSchema } from '@/schemas/referrals';

export const sessionSchema = z.object({
  id: z.number().int().positive(),
  patient_school_id: z.string(),
  counsellor_user_id: z.number().int().positive(),
  started_at: z.string(),
  ended_at: z.string().nullable(),
});
export type Session = z.infer<typeof sessionSchema>;

export const sessionDetailSchema = sessionSchema.extend({
  patient_display_name: z.string(),
  queue_entry_id: z.number().int().positive().nullable(),
  queue_number: z.string().nullable(),
  queue_status: z.enum(['waiting', 'called', 'in_session', 'done', 'skipped']).nullable(),
  purpose: z.string().nullable(),
  appointment_id: z.number().int().positive().nullable(),
  incoming_referral_id: z.number().int().positive().nullable(),
  note_count: z.number().int().min(0),
  outgoing_referral: referralSchema.nullable(),
});
export type SessionDetail = z.infer<typeof sessionDetailSchema>;

export const noteSchema = z.object({
  session_id: z.number().int().positive(),
  plaintext: z.string(),
  key_version: z.number().int(),
  created_at: z.string(),
  // Amendment chain (F15): id + optional pointer to the note this row
  // supersedes. Notes are insert-only, so corrections happen by amendment.
  id: z.number().int().positive().optional(),
  supersedes_note_id: z.number().int().positive().nullable().optional(),
});
export type Note = z.infer<typeof noteSchema>;

export const openSessionSchema = z.object({
  patient_school_id: z.string().min(1).max(32),
});
export type OpenSessionInput = z.infer<typeof openSessionSchema>;

export const writeNotesSchema = z.object({
  plaintext: z.string().min(1).max(16384),
  supersedes_note_id: z.number().int().positive().optional(),
});
export type WriteNotesInput = z.infer<typeof writeNotesSchema>;

File: mobile/lib/core/models/counselling.dart
/// Counselling sessions — mirrors `frontend/src/schemas/counselling.ts`
/// (`sessionSchema`).
library;

import 'referral.dart';

class CounsellingSession {
  CounsellingSession({
    required this.id,
    required this.patientSchoolId,
    required this.counsellorUserId,
    required this.startedAt,
    this.endedAt,
    this.patientDisplayName,
    this.queueEntryId,
    this.queueNumber,
    this.queueStatus,
    this.purpose,
    this.appointmentId,
    this.incomingReferralId,
    this.noteCount = 0,
    this.outgoingReferral,
  });

  factory CounsellingSession.fromJson(Map<String, dynamic> json) =>
      CounsellingSession(
        id: json['id'] as int,
        patientSchoolId: (json['patient_school_id'] ?? '') as String,
        counsellorUserId: (json['counsellor_user_id'] ?? 0) as int,
        startedAt: (json['started_at'] ?? '') as String,
        endedAt: json['ended_at'] as String?,
        patientDisplayName: json['patient_display_name'] as String?,
        queueEntryId: json['queue_entry_id'] as int?,
        queueNumber: json['queue_number'] as String?,
        queueStatus: json['queue_status'] as String?,
        purpose: json['purpose'] as String?,
        appointmentId: json['appointment_id'] as int?,
        incomingReferralId: json['incoming_referral_id'] as int?,
        noteCount: (json['note_count'] ?? 0) as int,
        outgoingReferral: json['outgoing_referral'] is Map<String, dynamic>
            ? Referral.fromJson(
                json['outgoing_referral'] as Map<String, dynamic>)
            : null,
      );

  final int id;
  final String patientSchoolId;
  final int counsellorUserId;
  final String startedAt;
  final String? endedAt;
  final String? patientDisplayName;
  final int? queueEntryId;
  final String? queueNumber;
  final String? queueStatus;
  final String? purpose;
  final int? appointmentId;
  final int? incomingReferralId;
  final int noteCount;
  final Referral? outgoingReferral;

  bool get isOpen => endedAt == null || endedAt!.isEmpty;
}

class CounsellingNote {
  CounsellingNote({
    required this.sessionId,
    required this.plaintext,
    required this.keyVersion,
    required this.createdAt,
  });

  factory CounsellingNote.fromJson(Map<String, dynamic> json) =>
      CounsellingNote(
        sessionId: (json['session_id'] ?? 0) as int,
        plaintext: (json['plaintext'] ?? '') as String,
        keyVersion: (json['key_version'] ?? 0) as int,
        createdAt: (json['created_at'] ?? '') as String,
      );

  final int sessionId;
  final String plaintext;
  final int keyVersion;
  final String createdAt;
}
